dbus/src/message/matchrule.rs

```diff
@@ -167,4 +167,300 @@ impl<'a> MatchRule<'a> {
+        fn key(&self) -> (&'a str, &'a str) {
+            let index = self.text.find('=').unwrap_or_else(|| self.text.len());
+
+            ((&self.text[..index]).trim(), &self.text[index + 1..])
+        }
+
+        /// Parses values as generic strings.
+        /// This does not do any validation (yet) with regards to supported characters.
+        fn value(&self) -> (&'a str, &'a str) {
+            let mut i = 0;
+            let mut quoted = false;
+            let mut escape = false;
+
+            for c in self.text.chars() {
+                match c {
+                    '\'' if !escape => {
+                        quoted = !quoted;
+                    }
+                    ',' if !quoted => {
+                        break;
+                    }
+                    '\\' if !quoted => {
+                        escape = true;
+                        i += 1;
+                        continue;
+                    }
+                    _ => {}
+                }
+                escape = false;
+
+                i += 1;
+            }
+
+            // Skip comma if there is still space in the buffer
+            if self.text.len() == i {
+                ((&self.text[..i]).trim(), &self.text[i..])
+            } else {
+                ((&self.text[..i]).trim(), &self.text[i + 1..])
+            }
+        }
+
+        /// Tokenizes a string into key-value-pairs
+        pub fn tokenize(&mut self) -> Result<TokenBuffer<'a>, Error> {
+            let mut rules = TokenBuffer::new();
+
+            while !self.text.is_empty() {
+                let (key, rest) = self.key();
+                self.text = rest;
+                let (value, rest) = self.value();
+                self.text = rest;
+                rules.try_push((key, value))
+                    .map_err(|_| Error::TooLong)?;
+            }
+            Ok(rules)
+        }
+    }
+
+    #[derive(Clone, Debug)]
+    /// Helper struct for parsing MatchRule's
+    pub struct Parser<'a> {
+        tokens: TokenBuffer<'a>,
+    }
+
+    impl<'a> Parser<'a> {
+        /// Builds a new parser after tokenizing the input string `text`.
+        pub fn new(text: &'a str) -> Result<Self, Error> {
+            Ok(Self {
+                tokens: Tokenizer::new(text).tokenize()?
+            })
+        }
+
+        /// Cleans a string from the string syntax allowed by DBus. This includes concatenating
+        /// things like `''\'''` to `'`. DBus strings sort of work like in a POSIX shell and
+        /// concatenation is implied. There is only one escape sequence and that is in an unquoted
+        /// substring a backslash may escape an ASCII apostrophe (U+0027).
+        /// This method is the only one within the parser that allocates and in theory could be
+        /// rewritten to taking a `&mut str` instead of returning a `String` since all
+        /// strings are |output| <= |buf.len()|.
+        fn clean_string(&self, buf: &str) -> String {
```

Comment:
I guess the easiest optimization here would be to have a clean_string check (this could be part of the `value` function if you like) that indicates whether or not the string actually needs to be cleaned; the common case would be that it does not need cleaning and so one could skip calling this function in the first place.

Reply:
Good point, I'll get on that. I've designed my parser with the worst case scenarios in mind e.g. `arg0=''\''',arg1='\',arg2=',',arg3='\\'` so I've written it in a rather defensive code style. I'm not all too happy with this method in general and the comment isn't quite truthful either since we'll usually be working with &'static str and not &'static mut str